[OCPNODE-553] Migrate ICSP to ImageDigest,TagMirrorSet CRDs

QiWang19 opened the pull request:
Migrate ICSP to ImageDigest,TagMirrorSet.
MCO has handled the object conversoin from icsp to ImageDiegstMirrorSet objects.

Epic: https://issues.redhat.com/browse/OCPNODE-521
Epic: https://issues.redhat.com/browse/OCPNODE-810

Signed-off-by: Qi Wang <qiwan@redhat.com>

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -71,55 +76,127 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
+// mergedDigestMirrorSets processes idmsRules and returns a set of ImageDigestMirrors, one for each Source value,
+// ordered consistently with the preference order of the individual entries (if possible)
+// E.g. given mirror sets (B, C) and (A, B), it will combine them into a single (A, B, C) set.
+func mergedDigestMirrorSets(idmsRules []*apicfgv1.ImageDigestMirrorSet) ([]apicfgv1.ImageDigestMirrors, error) {
```
At a first glance it seems to me that it should be possible to share more of the logic, but I haven’t read the code that carefully. (I do appreciate there’s a `topoSortRepos` now.)

_Maybe_ an object that maintains `disjointSets`/`mirrorBlockSource`, with methods very vaguely like `addMirrorSet(source, neverContactSource, mirrors)` and `mergedMirrorSets() []struct{source, neverContactSource, mirrors}`? That way the two functions here would deal _only_ with the data conversion with no logic at all.

Or _maybe_, instead of `mergedMirrorSets`, a `sourceList() []string` and `updateRegistry(source, *Registry)`, to share more of the code in `EditRegistriesConfig` that just consumes the data from `mergedMirrorSets`.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -71,55 +76,127 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 // A valid scope is in the form of registry/namespace...[/repo] (can also refer to sysregistriesv2.Registry.Prefix)
 // NOTE: Validation of wildcard entries is done before EditRegistriesConfig is called in the MCO code.
-func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScopes, blockedScopes []string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) error {
+func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScopes, blockedScopes []string, idmsRules []*apicfgv1.ImageDigestMirrorSet, itmsRules []*apicfgv1.ImageTagMirrorSet) error {
```
How is v1alpha1.ICSP going to be handled?

Considering there are so many users that are going to need updating to support he new CRs, I’d expect the ICSP handling to exist in the shared code (i.e. in this package), definitely not copy&pasted in every caller.

mtrmac reviewed (commented):
Just an _extremely brief_ skim for now, not really a review.

Let’s start with the ICSP discussion.

QiWang19 commented:
@mtrmac PTAL. More code got shared.

nee1esh commented:
/retest

QiWang19 commented:
/retest

QiWang19 committed:
Add support ImageDigest,TagMirrorSet CRDs

Add support for ImageDigestMirrorSet, ImageTagMirrorSet CRD

Epic: https://issues.redhat.com/browse/OCPNODE-521
Epic: https://issues.redhat.com/browse/OCPNODE-810

Signed-off-by: Qi Wang <qiwan@redhat.com>

QiWang19 commented:
@mtrmac could you review it? ICSP object migration should be done in MCO, this PR update uses new APIs.

rphillips commented:
/remove-lifecycle rotten

rphillips commented:
The pointer indirection is likely not needed. Can we refactor this by removing the pointer indirection such that the functions do not take a pointer, like:

`		disjointSets:      map[string]*[]*[]apicfgv1.ImageMirror{},`

and

```
func (sets *mirrorSets) addMirrorSets(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors *[]apicfgv1.ImageMirror)
```

The mirrors attribute probably doesn't need to be a pointer. Likewise for the other functions in the file.

mtrmac commented:
> @mtrmac could you review it? ICSP object migration should be done in MCO, this PR update uses new APIs.

Is that viable?
- https://github.com/openshift/enhancements/pull/873/files still says the v1alpha1 API will remain. So external users are going to be creating them over the lifetime of the cluster.
- At the very least https://github.com/mtrmac/openshift-apiserver/blob/master/pkg/image/apiserver/registry/imagestreamimport/rest.go#L217-L221 calls EditRegistries (and per the IDMS/ITMS enhancement, IIRC quite a few more repositories need to do so in the future). So is the idea that MCO is going to immediately convert, and delete, every created ICSP , and API consumers can then only deal with the new types? That seems fine for the controllers…
- … but what about users? They are going to create a ICSP and just not see it in the API after a few seconds. Is that reasonable?

(It might well be fine. It’s just not what I vaguely remember the earlier migration conversations to be like, and I just thought I’d ask before we commit to that direction.)

mtrmac commented:
> So is the idea that MCO is going to immediately convert, and delete, every created ICSP , and API consumers can then only deal with the new types?

I’m sorry, actually reading the code, MCO is keeping the ICSP resources around, so they are still available to old consumers.

That seems perfect to me. Back to actually reading this PR…

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
-	for _, mirror := range set.Mirrors {
-		if mirror != set.Source {
+// rdmContainsARealMirror returns true if mirrors contains at least one entry that is not source.
+func rdmContainsARealMirror(source string, mirrors []apicfgv1.ImageMirror) bool {
```
`rdm` refers to `RepositoryDigestMirrors`, so this should probably be renamed. `mirrorsContainARealMirror`? `mirrorsContainSomeOtherThan`?

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	}
+}
+
+// sourceList collects the mirror sources and sort them in increasing order
```
```suggestion
// sourceList collects the mirror sources and returns them sorted in increasing order
```
to be a bit clearer that this this does not affect the `mirrorSets` object.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+
+func (sets *mirrorSets) addMirrorSets(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors []apicfgv1.ImageMirror) {
+	if !rdmContainsARealMirror(source, mirrors) {
+		return // No mirrors (or mirrors that only repeat the authoritative source) is not really a mirror set.
```
Is it possible to have such a no-mirror object with `NeverContactSource`? If so, should we still set `mirrorBlockSource`?

(I’m most tempted to say that such an object is invalid and should be rejected, I’m not sure if it is practical. If it isn’t, both setting `mirrorBlockSource` and the current approach is, I guess, acceptable, but with the current approach, a extending the comment that yes, we ignore `mirrorSourcePolicy` and that’s intentional, would be useful.)

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+		res = append(res, imageTagMirror)
 	}
-	sort.Strings(sources)
+	return res, nil
```
I don’t think we _need_ to return correctly-typed `Image{Tag,Digest}Mirrors` from the pair of these functions — so we could define a local `type mergedMirrorsItem struct { source, mirrors, mirrorSourcePolicy }` and all of this loop could be also shared.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
 	return res, nil
 }
 
+func topoSortRepos(source string, ds *[][]apicfgv1.ImageMirror) ([]string, error) {
```
This is not just topological sort. Maybe `{shared,raw}MergedMirrors`? (Or, alternatively, if the loop calling this can be made shared, it can be inlined back. OTOH having this code as a separate function might be a better granularity for unit testing.)

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,208 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+			}
+		}
+		if !sourceInGraph {
+			// The build of mirrorSets guarantees len(set.Mirrors) > 0.
```
```suggestion
			// mirrorSets.addMirrorSets guarantees len(set.Mirrors) > 0.
```
or, alternatively, make this, or the loop calling this, a method on `mirrorSets`.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -278,3 +463,16 @@ func IsValidRegistriesConfScope(scope string) bool {
+
+// RejectMultiUpdateMirrorSetObjs returns error if icsp objects exist with imagedigestmirrorset or imagetagmirrorset objects
+// to avoid existing mirror settings get updated by others
+func RejectMultiUpdateMirrorSetObjs(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy,
```
So the new plan is again not to have a controller sync between the old/new objects, and for every consumer to read both the old and new values?

(That’s fine, I’m just trying to keep up.)

mtrmac replied:
Oh, you have actually confirmed exactly that elsewhere already.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -278,3 +463,16 @@ func IsValidRegistriesConfScope(scope string) bool {
+
+// RejectMultiUpdateMirrorSetObjs returns error if icsp objects exist with imagedigestmirrorset or imagetagmirrorset objects
+// to avoid existing mirror settings get updated by others
+func RejectMultiUpdateMirrorSetObjs(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy,
```
Do we anticipate having other external callers? If not, can this be made private?

QiWang19 replied:
MCO can call this at syncImageConfig() and RunImageBootstrap(). I thought this might help reject at an early stage.

mtrmac replied:
ACK, let’s keep this public then.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -187,16 +319,31 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
+	for _, mirrorSet := range icspMirrorSets {
+		reg := getRegistryEntry(mirrorSet.Source)
+		updateRegistry(mirrorSet.Mirrors, mirrorSet.MirrorSourcePolicy, sysregistriesv2.MirrorByDigestOnly, reg)
+	}
```
With the `mergedMirrorItem` type discussed elsewhere, some variant of the `updateRegistry` function could contain the whole loop.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -238,7 +385,46 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
+			// So, if there is any mirror defined for a more specific sub-scope of mirrorSet.Source,
+			// it must already exist with non-empty reg.Mirrors.
+			if scope != mirroredScope && ScopeIsNestedInsideScope(scope, mirroredScope) && len(reg.Mirrors) == 0 {
+				updated, err := mirrorsAdjustedForNestedScope(mirroredScope, scope, mirroredReg.Mirrors)
```
A test case for this in the `insecure+blocked scopes inside a configured mirror` situation would be nice.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -248,7 +434,6 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
 			// it must already exist with non-empty reg.Mirrors.
 			if scope != mirroredScope && ScopeIsNestedInsideScope(scope, mirroredScope) && len(reg.Mirrors) == 0 {
-				reg.MirrorByDigestOnly = mirroredReg.MirrorByDigestOnly
 				updated, err := mirrorsAdjustedForNestedScope(mirroredScope, scope, mirroredReg.Mirrors)
```
A test case for this in the `insecure+blocked scopes inside a configured mirror` situation would be nice.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -238,7 +385,46 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
 	}
-	for _, mirrorSet := range mirrorSets {
+
+	for _, mirrorSet := range icspMirrorSets {
```
Could the three copies be consolidated? With `mergedMirrorItem`, we could just append the three `*MirrorSets` arrays together and there could be a single loop.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -80,38 +81,44 @@ func TestRDMContainsARealMirror(t *testing.T) {
-			set := apioperatorsv1alpha1.RepositoryDigestMirrors{
-				Source:  source,
-				Mirrors: tt.mirrors,
+			set := &apicfgv1.ImageTagMirrorSet{
```
Non-blocking: If `set` is not used other than to read one of its fields, it doesn’t need to be constructed.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -219,7 +226,261 @@ func TestMergedMirrorSets(t *testing.T) {
+	}
+}
+
+func TestMergedTagMirrorSets(t *testing.T) {
```
If the tests are copy&pasted, could they instead be moved into a single global variable, with the tests inner loops making any type conversions that are necessary?

(All of this might get simpler if most of `the mergedMirrorSets` core logic can be shared.)

QiWang19 replied:
fixed

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -497,4 +1021,117 @@ func TestEditRegistriesConfig(t *testing.T) {
+			_, err := toml.Decode(string(templateBytes), &config)
+			require.NoError(t, err)
+			err = EditRegistriesConfig(&config, tt.insecure, tt.blocked, tt.icspRules, tt.idmsRules, tt.itmsRules)
+			require.NotNil(t, err)
```
Absolutely non-blocking: `assert.NoError` would be a better match here.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -497,4 +1021,117 @@ func TestEditRegistriesConfig(t *testing.T) {
+		idmsRules         []*apicfgv1.ImageDigestMirrorSet
+		itmsRules         []*apicfgv1.ImageTagMirrorSet
+		icspRules         []*apioperatorsv1alpha1.ImageContentSourcePolicy
+		want              sysregistriesv2.V2RegistriesConf
```
`insecure`, `blocked`, `want` are unused, so that can be dropped.

Non-blocking: This test could be shorter and somewhat easier to fit on the screen with:
- Only including one set of mirrors instead of three in every item.
- Making the test data `idmsNonempty, itmsNonempty, icspNonempty bool`, and letting the testing loop provide data or an empty array.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -408,16 +757,195 @@ func TestEditRegistriesConfig(t *testing.T) {
+			name: "imageTagMirrorSet",
+			itmsRules: []*apicfgv1.ImageTagMirrorSet{
+				{
+					Spec: apicfgv1.ImageTagMirrorSetSpec{
+						ImageTagMirrors: []apicfgv1.ImageTagMirrors{ // other.com is neither insecure nor blocked
+							{Source: "insecure.com/ns-i1", Mirrors: []apicfgv1.ImageMirror{"blocked.com/ns-b1", "other.com/ns-o1"}},
+							{Source: "blocked.com/ns-b/ns2-b", Mirrors: []apicfgv1.ImageMirror{"other.com/ns-o2", "insecure.com/ns-i2"}},
+							{Source: "other.com/ns-o3", Mirrors: []apicfgv1.ImageMirror{"insecure.com/ns-i2", "blocked.com/ns-b/ns3-b", "foo.insecure-example.com/bar"}},
+						},
```
This test does not set anything to insecure nor blocked, so using the `insecure.com`/`blocked.com` domains is a bit unexpected.

What is being tested here? Just a smoke-test of correctly creating tag entries? Two entries would be enough for that.

mtrmac reviewed (commented):
Two API semantics questions

And I’d like to push a bit further on the code sharing: The need to propagate blocked/insecure to sub-scopes is somewhat non-obvious, and having that triplicated feels a bit risky to me.

mtrmac commented:
The failing tests do seem to suggest that updating to Go ≥ 1.18 is required to be able to consume the new API module.

mtrmac commented:
> The failing tests do seem to suggest that updating to Go ≥ 1.18 is required to be able to consume the new API module.

Ah, that’s being handled in https://github.com/openshift/release/pull/31916 .

QiWang19 commented:
@mtrmac PTAL

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
 
-	// Sort the sets of mirrors by Source to ensure deterministic output
+// sourceList collects the mirror sources and returns them sorted in increasing order
+func (sets *mirrorSets) sourceList() []string {
```
This now has only one caller and it can be inlined back.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	return sources
+}
+
+func (sets *mirrorSets) addMirrorSets(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors []apicfgv1.ImageMirror) {
```
```suggestion
func (sets *mirrorSets) addMirrorSet(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors []apicfgv1.ImageMirror) {
```
`sets` is multiple sets, this is adding a single one.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	if !mirrorsContainsARealMirror(source, mirrors) {
+		return // No mirrors (or mirrors that only repeat the authoritative source) is not really a mirror set. Ignore mirrorSourcePolicy intentionally.
+	}
+	strMirrors := []string{}
+	for _, m := range mirrors {
+		strMirrors = append(strMirrors, (string(m)))
+	}
```
Absolutely non-blocking: This copy is probably avoidable, perhaps with some more casts in the merging implementation.

(It’s a bit weird for ICSP to do a copy to convert to `ImageMirror`, only for this code to convert back to `strings`.)

But the cost is trivial either way, I’m not worried about it at all.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+}
+
+// mergedMirrors generates deterministic order of mirrors for a given source
+func (sets *mirrorSets) mergedMirrors(source string) (string, []string, error) {
```
This doesn’t need to return `source`, it’s just the caller-supplied value.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	return source, sortedRepos, nil
+}
+
+type mergedMirrorsItem struct {
```
Reading the full context, (contrary to my earlier suggestion — my mistake) `mergedMirrorSet` is probably the right name. That’s what it is (unlike “item”, which says nothing), and both functions like `mergedICSPMirrorSets` and variables like `tagMirrorSets` use that naming.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+type mergedMirrorsItem struct {
+	source             string
+	mirrors            []string
+	mirrorSourcePolicy apicfgv1.MirrorSourcePolicy
```
Absolutely non-blocking: This could be a `blockSource bool`, to remove the need to worry about unknown string values. OTOH this string might be more useful in the test data.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+}
+
+// mergedMirrorsItems converts the set of mirrors to slice of mergedMirrorsItem
+func mergedMirrorsItems(sets *mirrorSets) ([]mergedMirrorsItem, error) {
```
```suggestion
func (sets *mirrorSets) mergedMirrorSets() ([]mergedMirrorSet, error) {
```

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+// ordered consistently with the preference order of the individual entries (if possible)
+// E.g. given mirror sets (B, C) and (A, B), it will combine them into a single (A, B, C) set.
+func mergedDigestMirrorSets(idmsRules []*apicfgv1.ImageDigestMirrorSet) ([]mergedMirrorsItem, error) {
+
```
(Drop this line)

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	return mergedMirrorsItems(repositoryMirrorSets)
+}
+
+func sharedMergedMirrors(source string, ds *[][]string) ([]string, error) {
```
This now has a single caller, and can be inlined back.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	return sortedRepos, nil
+}
+
+func updateRegistry(mergedMirrorsItems []mergedMirrorsItem, pullFromMirror string, getRegistryEntry func(scope string) *sysregistriesv2.Registry) {
```
The need to pass `getRegistryEntry` in is a bit awkward.

Maybe move this inside `EditRegistriesConfig`, as another nested function.

(Eventually we might need to split `EditRegistriesConfig` into a data structure + methods, but I don‘t think that’s immediately urgent. A smaller diff is also valuable.)

mtrmac replied:
```suggestion
func {addMirrorsToRegistries,configureMirrors}(mergedMirrorsItems []mergedMirrorsItem, pullFromMirror string, getRegistryEntry func(scope string) *sysregistriesv2.Registry) {
```

`updateRegistry` does not say much about what is being updated.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -238,8 +353,10 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
 		}
 	}
+
+	mirrorSets := append(icspMirrorSets, append(digestMirrorSets, tagMirrorSets...)...)
```
Non-blocking: 
```suggestion
	allMirrorSets := append(icspMirrorSets, append(digestMirrorSets, tagMirrorSets...)...)
```
to be a bit more explicit here

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -76,155 +77,166 @@ func TestIsValidRegistriesConfScope(t *testing.T) {
+						imgMirrors = append(imgMirrors, apicfgv1.ImageMirror(m))
+					}
+					itmElement := apicfgv1.ImageTagMirrors{Source: item.source, Mirrors: imgMirrors}
+					if tc.name == "Separate mirror sets" && item.source == "source.example.net" {
```
Please use `item.mirrorSourcePolicy` for this, so that it is clearly visible in the test data. (It’s fine for it to be default-set to `""` in most entries.)

ICSP can then skip the test that exercises this if it can’t represent it.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -76,155 +77,166 @@ func TestIsValidRegistriesConfScope(t *testing.T) {
+				})
+			}
+			tcRes := tc.result
+			if tc.name == "Separate mirror sets" {
```
Please do this in the data.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -76,155 +77,166 @@ func TestIsValidRegistriesConfScope(t *testing.T) {
+			if err != nil {
+				t.Errorf("Error %v", err)
+				return
+			}
```
Absolutely non-blocking: `require.NoError(t, err)` is shorter. This works just fine — and I realize the original was this way as well.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -408,10 +508,231 @@ func TestEditRegistriesConfig(t *testing.T) {
+			idmsRules: []*apicfgv1.ImageDigestMirrorSet{
+				{
+					Spec: apicfgv1.ImageDigestMirrorSetSpec{
+						ImageDigestMirrors: []apicfgv1.ImageDigestMirrors{ // other.com is neither insecure nor blocked
```
This comment seems inapplicable.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -408,10 +508,231 @@ func TestEditRegistriesConfig(t *testing.T) {
+			itmsRules: []*apicfgv1.ImageTagMirrorSet{
+				{
+					Spec: apicfgv1.ImageTagMirrorSetSpec{
+						ImageTagMirrors: []apicfgv1.ImageTagMirrors{ // other.com is neither insecure nor blocked
```
This comment seems inapplicable.

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -76,155 +77,166 @@ func TestIsValidRegistriesConfScope(t *testing.T) {
-				{Source: "source.example.net", Mirrors: []string{"z1.example.net", "y2.example.net", "x3.example.net"}},
-			},
-		},
-		{
```
The following test cases were dropped. Is there a specific reason for that? If not, please add them back.

QiWang19 replied:
Added back. Accidently removed during the last code cleanup.

mtrmac reviewed (commented):
Thanks!

Implementation LGTM, just a few cleanup suggestions. Looking at the tests, I’m unsure why various tests were just dropped.

QiWang19 commented:
@mtrmac PTAL.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,162 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+		}
+		sourceInGraph := false
+		for _, m := range mirrors {
+			if string(m) == source {
```
Non-blocking: The `string()` cast can now be dropped.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -35,81 +36,162 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+func mergedTagMirrorSets(itmsRules []*apicfgv1.ImageTagMirrorSet) ([]mergedMirrorSet, error) {
+	tagMirrorSets := newMirrorSets()
+	for _, itms := range itmsRules {
+		for i := range itms.Spec.ImageTagMirrors {
+			set := itms.Spec.ImageTagMirrors[i]
```
Absolutely non-blocking:
```go
for _, set := range … { /* use set */ }
```
would be one variable and one line shorter. (In all three instances.)

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -248,14 +355,14 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
 				reg.Mirrors = updated
 			}
 		}
+
```
(Absolutely non-blocking: drop this line? I don’t think it separates “paragraphs” / individual ideas when placed here.)

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -76,155 +77,229 @@ func TestIsValidRegistriesConfScope(t *testing.T) {
+func TestMergedICSPMirrorSets(t *testing.T) {
+	for _, tc := range mergedMirrorsetsTestcases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.name == "Separate mirror sets with mirrorSourcePolicy set" {
```
Please make this driven by a real data field.

Either here,
```go
if tc.requiresMirrorSourcePolicy { t.Skip() }
```
(and set the `tc.requiresMirrorSourcePolicy` field on the test case), or (preferably?) make this fully automatic below, with something like

```
if item.mirrorSourcePolicy != "" {
    t.Skip("…")
}
```

mtrmac reviewed (commented):
One very last thing - the `tc.name == …` condition.

LGTM otherwise.

QiWang19 committed:
bump c/image

Signed-off-by: Qi Wang <qiwan@redhat.com>

mtrmac commented:
/approve

mtrmac commented:
/lgtm

mtrmac commented:
Thanks!
